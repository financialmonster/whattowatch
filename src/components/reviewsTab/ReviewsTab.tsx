import React, { FC, memo } from 'react';
import { List } from 'immutable';

import { useFetchReviews } from 'hooks/useFetchReviews';
import { Spinner } from 'components/spinner/Spinner';
import { isReviews }  from 'utils';
import { Review } from 'components/review/Review';
import { Message } from 'components/message/Message';

type TReviewsTabProps = {
    id: number;
}

export const ReviewsTab: FC<TReviewsTabProps> = memo(({ id }) => {
    const {isReviewsFetching, reviewsError, reviews} = useFetchReviews(id);

    if(isReviewsFetching) {
        return <Spinner />;
    }

    if(reviewsError) {
        return (
            <Message>
                {(reviewsError as Error).message}
            </Message>
        );
    }

    let reviewsJSX: List<JSX.Element> | undefined;
    let leftColReviewsJSX: List<JSX.Element> | undefined;
    let rightColReviewsJSX: List<JSX.Element> | undefined;

    if(isReviews(reviews)) {
        reviewsJSX = reviews.sort((a, b) => Date.parse(b.get(`date`)) - Date.parse(a.get(`date`)))
                            .map((review) => <Review review={review} key={review.get(`id`)} />);

        leftColReviewsJSX = reviewsJSX.filter((_, ind) => !(ind % 2));
        rightColReviewsJSX = reviewsJSX.filter((_, ind) => ind % 2);
    }
    
    return (
        <div className = "movie-card__reviews movie-card__row">
            {(reviewsJSX) &&
                <>
                    <div className="movie-card__reviews-col">
                        {leftColReviewsJSX}
                    </div>
                    <div className="movie-card__reviews-col">
                        {rightColReviewsJSX}
                    </div>
                </>
            }
            {(!reviewsJSX) && 
                <Message>
                    No reviews yet. Your will be the first!
                </Message>
            }
        </div>
    );
});
