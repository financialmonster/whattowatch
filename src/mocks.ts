import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
