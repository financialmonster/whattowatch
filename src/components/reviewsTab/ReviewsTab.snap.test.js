import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { ReviewsTab } from './ReviewsTab'; 
import { createStoreMock } from 'mocks';

it(`ReviewsTab: correctly rendered`, () => {
    const mockStore = configureStore();
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <ReviewsTab />
        </Provider> 
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
