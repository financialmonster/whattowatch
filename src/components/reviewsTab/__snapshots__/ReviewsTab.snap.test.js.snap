// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsTab: correctly rendered 1`] = `
<div
  className="sc-AxjAm fcmpGc"
>
  <div
    className="sc-AxirZ fLbCRQ"
    rotate="rotate(0deg)"
  />
  <div
    className="sc-AxirZ iRnakP"
    rotate="rotate(30deg)"
  />
  <div
    className="sc-AxirZ jJDLSA"
    rotate="rotate(60deg)"
  />
  <div
    className="sc-AxirZ cfISCN"
    rotate="rotate(90deg)"
  />
  <div
    className="sc-AxirZ jYaMud"
    rotate="rotate(120deg)"
  />
  <div
    className="sc-AxirZ kCauMN"
    rotate="rotate(150deg)"
  />
  <div
    className="sc-AxirZ bbuhJx"
    rotate="rotate(180deg)"
  />
  <div
    className="sc-AxirZ RfLVX"
    rotate="rotate(210deg)"
  />
  <div
    className="sc-AxirZ iDvjl"
    rotate="rotate(240deg)"
  />
  <div
    className="sc-AxirZ fchBdo"
    rotate="rotate(270deg)"
  />
  <div
    className="sc-AxirZ fGeryM"
    rotate="rotate(300deg)"
  />
  <div
    className="sc-AxirZ eVSKan"
    rotate="rotate(330deg)"
  />
</div>
`;
